fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseRunnerImpl.java

```diff
@@ -372,7 +342,66 @@ public T run(@Nonnull Function<? super FDBRecordContext, ? extends T> retriable)
     public <T> CompletableFuture<T> runAsync(@Nonnull final Function<? super FDBRecordContext, CompletableFuture<? extends T>> retriable,
                                              @Nonnull final BiFunction<? super T, Throwable, ? extends Pair<? extends T, ? extends Throwable>> handlePostTransaction,
                                              @Nullable List<Object> additionalLogMessageKeyValues) {
-        return new RunRetriable<T>(additionalLogMessageKeyValues).runAsync(retriable, handlePostTransaction);
+        final RetryTask.Builder<T> builder = new RetryTask.Builder<>(() -> {
+            FDBRecordContext ctx = openContext();
+            return retriable.apply(ctx).thenCompose(val ->
+                    ctx.commitAsync().thenApply(vignore -> val)
+            ).handle((result, ex) -> {
+                Pair<? extends T, ? extends Throwable> newResult = handlePostTransaction.apply(result, ex);
+                ctx.close();
+                if (newResult.getRight() == null) {
+                    return newResult.getLeft();
+                } else {
+                    RuntimeException runtimeException = database.mapAsyncToSyncException(newResult.getRight());
+                    throw runtimeException;
+                }
+            });
+        }, getMaxAttempts());
+
+        builder.setPossiblyRetry(states -> {
+            if (closed) {
+                // Outermost future should be cancelled, but be sure that this doesn't appear to be successful.
+                states.possibleException.set(new RunnerClosed());
```

Comment:
It kind of seems like `RetryTask` should handle making sure that tasks stop running when the runner is closed. Or put another way, I'd note that the `OnlineIndexer` uses `.close()` in exactly the same way, which is a sign that it's something common to retry things rather than peculiar to the `FDBDatabaseRunner` (whereas, say, retrying only certain exceptions is something that the `FDBDatabaseRunner` in particular would care about).

I think this might mean that this class needs, like, a "runners to close" list or something, and then when this class is `close`d, it also calls close on the runner.